presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/RemoveUnsupportedDynamicFilters.java

```diff
@@ -0,0 +1,178 @@
+
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static io.prestosql.sql.DynamicFilters.extractDynamicFilters;
+import static io.prestosql.sql.DynamicFilters.getDescriptor;
+import static io.prestosql.sql.ExpressionUtils.combineConjuncts;
+import static io.prestosql.sql.ExpressionUtils.extractConjuncts;
+import static io.prestosql.sql.tree.BooleanLiteral.TRUE_LITERAL;
+
+/**
+ * Dynamic filters are supported only right after TableScan and only if the subtree is on the probe side of some downstream join node
+ * Dynamic filters are removed from JoinNode if there is no consumer for it on probe side
+ */
+public class RemoveUnsupportedDynamicFilters
+        implements PlanOptimizer
+{
+    private static class SupportedDynamicFiltersContext
+    {
+        private final ImmutableSet<String> allowedDynamicFilterIds;
+        private final Set<String> consumedDynamicFilterIds;
+
+        SupportedDynamicFiltersContext(ImmutableSet<String> allowedDynamicFilterIds, Set<String> consumedDynamicFilterIds)
+        {
+            this.allowedDynamicFilterIds = allowedDynamicFilterIds;
+            this.consumedDynamicFilterIds = consumedDynamicFilterIds;
+        }
+
+        ImmutableSet<String> getAllowedDynamicFilterIds()
+        {
+            return allowedDynamicFilterIds;
+        }
+
+        Set<String> getConsumedDynamicFilterIds()
+        {
+            return consumedDynamicFilterIds;
+        }
+    }
+
+    @Override
+    public PlanNode optimize(PlanNode plan, Session session, TypeProvider types, SymbolAllocator symbolAllocator, PlanNodeIdAllocator idAllocator, WarningCollector warningCollector)
+    {
+        return SimplePlanRewriter.rewriteWith(new RemoveUnsupportedDynamicFilters.Rewriter(), plan, new SupportedDynamicFiltersContext(ImmutableSet.of(), new HashSet<>()));
+    }
+
+    private class Rewriter
+            extends SimplePlanRewriter<SupportedDynamicFiltersContext>
+    {
+        @Override
+        public PlanNode visitJoin(JoinNode node, RewriteContext<SupportedDynamicFiltersContext> context)
+        {
+            ImmutableSet<String> allowedDynamicFilterIds = ImmutableSet.<String>builder()
+                    .addAll(node.getDynamicFilters().keySet())
+                    .addAll(context.get().getAllowedDynamicFilterIds())
+                    .build();
+
+            SupportedDynamicFiltersContext leftContext = new SupportedDynamicFiltersContext(allowedDynamicFilterIds, new HashSet<>());
+            PlanNode left = context.rewrite(node.getLeft(), leftContext);
+            Set<String> consumedProbeSide = leftContext.getConsumedDynamicFilterIds();
+            Map<String, Symbol> dynamicFilters = node.getDynamicFilters().entrySet().stream()
+                    .filter(entry -> consumedProbeSide.contains(entry.getKey()))
```

Comment:
Yes, my intention is to preserve dynamic filters which were consumed on probe side in join node. I am attempting to do this by creating a new JoinNode using the `dynamicFilters` map created here if there is any change from the `dynamicFilters` map within current join node. JoinNode::dynamicFilters is immutable so it can't be modified in-place.
Should this be done differently ?

Reply:
> Yes, my intention is to preserve dynamic filters which were consumed on probe side in join node. 

+1. That's correct. Skip my comment